File: src/symeraseme/core/deadlines.py
"""Tick engine — proactive lifecycle management for removal requests."""

from __future__ import annotations

import logging
import sqlite3
from dataclasses import dataclass, field
from datetime import UTC, datetime, timedelta
from typing import Any

from symeraseme.core.datetime_utils import parse_iso_datetime as _parse_dt

logger = logging.getLogger(__name__)

JURISDICTION_DEADLINES: dict[str, int] = {
    "GDPR": 30,
    "CCPA": 45,
    "CPRA": 45,
    "LGPD": 30,
    "PIPEDA": 30,
}

REMINDER_DAYS = 7
DPA_ESCALATION_DAYS = 14
RE_SCAN_DAYS = 90


@dataclass
class TickAction:
    request_id: int
    broker_id: str
    campaign_id: str
    current_status: str
    action_type: str
    event_type: str
    description: str
    payload: dict[str, Any] = field(default_factory=dict)
    dry_run: bool = False


def run_tick(
    *,
    dry_run: bool = False,
    reference_date: datetime | None = None,
    batch_size: int | None = None,
) -> list[TickAction]:
    """Run one tick cycle.

    Scans all removal requests and produces actions for those
    whose next_action_at <= reference_date (or now).
    """
    actions: list[TickAction] = []
    now = reference_date or datetime.now(UTC)

    from symeraseme.core.repositories.deadlines import fetch_tick_candidates

    rows = fetch_tick_candidates(now.isoformat(), batch_size=batch_size)

    for row in rows:
        req = dict(row)
        status = req.get("current_status", "PLANNED")
        actions.extend(_tick_for_request(req, status, now, dry_run=dry_run))

    return actions


def _tick_for_request(
    req: dict[str, Any],
    status: str,
    now: datetime,
    *,
    dry_run: bool = False,
) -> list[TickAction]:
    actions: list[TickAction] = []
    rid = req["id"]

    sent_at = _parse_dt(req.get("sent_at"))
    deadline_at = _parse_dt(req.get("deadline_at"))
    resolved_at = _parse_dt(req.get("resolved_at"))
    reminders_sent = req.get("reminders_sent", 0)
    escalation_level = req.get("escalation_level", 0)
    jurisdiction = req.get("jurisdiction", "GDPR")

    deadline_days = JURISDICTION_DEADLINES.get(jurisdiction, 30)

    if status == "AWAITING_ACK":
        action = _check_reminder(rid, req, sent_at, now, reminders_sent, dry_run=dry_run)
        if action:
            actions.append(action)

    elif status == "AWAITING_RESPONSE":
        action = _check_deadline(
            rid, req, sent_at, deadline_at, deadline_days, now, dry_run=dry_run
        )
        if action:
            actions.append(action)

    elif status == "OVERDUE":
        action = _check_dpa_escalation(
            rid, req, deadline_at, now, escalation_level, dry_run=dry_run
        )
        if action:
            actions.append(action)

    elif status == "CONFIRMED":
        action = _check_rescan(rid, req, resolved_at, now, dry_run=dry_run)
        if action:
            actions.append(action)

    return actions


def _check_reminder(
    rid: int,
    req: dict[str, Any],
    sent_at: datetime | None,
    now: datetime,
    reminders_sent: int,
    *,
    dry_run: bool = False,
) -> TickAction | None:
    """Send reminder if AWAITING_ACK for more than REMINDER_DAYS."""
    if sent_at is None:
        return None

    days_elapsed = (now - sent_at).days
    if days_elapsed < REMINDER_DAYS:
        return None

    # Exponential backoff: next reminder at 2^n * REMINDER_DAYS
    next_threshold = REMINDER_DAYS * (2**reminders_sent)
    if days_elapsed < next_threshold:
        return None

    return TickAction(
        request_id=rid,
        broker_id=req.get("broker_id", ""),
        campaign_id=req.get("campaign_id", ""),
        current_status="AWAITING_ACK",
        action_type="send_reminder",
        event_type="REMINDER_SENT",
        description=f"Send reminder #{reminders_sent + 1} ({days_elapsed}d since sent)",
        payload={"days_since_sent": days_elapsed, "count": reminders_sent + 1},
        dry_run=dry_run,
    )


def _check_deadline(
    rid: int,
    req: dict[str, Any],
    sent_at: datetime | None,
    deadline_at: datetime | None,
    deadline_days: int,
    now: datetime,
    *,
    dry_run: bool = False,
) -> TickAction | None:
    """Mark request as overdue if deadline has passed."""
    if deadline_at is not None and now >= deadline_at:
        return TickAction(
            request_id=rid,
            broker_id=req.get("broker_id", ""),
            campaign_id=req.get("campaign_id", ""),
            current_status="AWAITING_RESPONSE",
            action_type="mark_overdue",
            event_type="DEADLINE_REACHED",
            description=f"Deadline reached ({deadline_days}d, passed {now - deadline_at})",
            payload={
                "deadline_days": deadline_days,
                "deadline_at": deadline_at.isoformat(),
            },
            dry_run=dry_run,
        )

    if sent_at and deadline_at is None:
        effective_deadline = sent_at + timedelta(days=deadline_days)
        if now >= effective_deadline:
            return TickAction(
                request_id=rid,
                broker_id=req.get("broker_id", ""),
                campaign_id=req.get("campaign_id", ""),
                current_status="AWAITING_RESPONSE",
                action_type="mark_overdue",
                event_type="DEADLINE_REACHED",
                description=f"Deadline reached ({deadline_days}d from sent)",
                payload={
                    "deadline_days": deadline_days,
                    "deadline_at": effective_deadline.isoformat(),
                },
                dry_run=dry_run,
            )

    return None


def _check_dpa_escalation(
    rid: int,
    req: dict[str, Any],
    deadline_at: datetime | None,
    now: datetime,
    escalation_level: int,
    *,
    dry_run: bool = False,
) -> TickAction | None:
    """Generate DPA complaint after 14 days in OVERDUE."""
    if escalation_level >= 2:
        return None

    if deadline_at is None:
        return None

    days_since_deadline = (now - deadline_at).days
    if days_since_deadline < DPA_ESCALATION_DAYS:
        return None

    return TickAction(
        request_id=rid,
        broker_id=req.get("broker_id", ""),
        campaign_id=req.get("campaign_id", ""),
        current_status="OVERDUE",
        action_type="draft_dpa_complaint",
        event_type="DPA_COMPLAINT_DRAFTED",
        description=f"DPA complaint ready ({days_since_deadline}d since deadline)",
        payload={"days_since_deadline": days_since_deadline},
        dry_run=dry_run,
    )


def _check_rescan(
    rid: int,
    req: dict[str, Any],
    resolved_at: datetime | None,
    now: datetime,
    *,
    dry_run: bool = False,
) -> TickAction | None:
    """Trigger re-scan after 90 days from confirmation."""
    if resolved_at is None:
        return None

    days_since_resolved = (now - resolved_at).days
    if days_since_resolved < RE_SCAN_DAYS:
        return None

    return TickAction(
        request_id=rid,
        broker_id=req.get("broker_id", ""),
        campaign_id=req.get("campaign_id", ""),
        current_status="CONFIRMED",
        action_type="trigger_rescan",
        event_type="RE_SCAN_TRIGGERED",
        description=f"Re-scan due ({days_since_resolved}d since resolution)",
        payload={"days_since_resolved": days_since_resolved},
        dry_run=dry_run,
    )


def apply_tick_actions(
    actions: list[TickAction],
) -> list[dict[str, Any]]:
    """Apply tick actions to the event store."""
    if not actions:
        return []

    from symeraseme.core.projection import rebuild_all_states

    results: list[dict[str, Any]] = []
    for action in actions:
        if action.dry_run:
            results.append(
                {
                    "request_id": action.request_id,
                    "action": action.action_type,
                    "event_type": action.event_type,
                    "description": action.description,
                    "executed": False,
                    "dry_run": True,
                }
            )
            continue

        try:
            from symeraseme.core.projection import append_event_and_project

            append_event_and_project(
                action.request_id,
                action.event_type,
                payload=action.payload,
                source="scheduler",
            )
            results.append(
                {
                    "request_id": action.request_id,
                    "action": action.action_type,
                    "event_type": action.event_type,
                    "description": action.description,
                    "executed": True,
                    "dry_run": False,
                }
            )
        except (sqlite3.Error, ValueError, RuntimeError) as e:
            logger.error("Failed to apply tick action %s: %s", action, e)
            results.append(
                {
                    "request_id": action.request_id,
                    "action": action.action_type,
                    "event_type": action.event_type,
                    "description": action.description,
                    "executed": False,
                    "error": str(e),
                }
            )

    # Batch-rebuild all states in a single O(1) pass
    rebuild_all_states()

    return results

File: internal/deadlines/deadlines.go
// Package deadlines is the Go port of src/symeraseme/core/deadlines.py —
// the tick engine that drives proactive lifecycle management: due-date
// checks per jurisdiction, reminder backoff, DPA escalation and periodic
// re-scans.
package deadlines

import (
	"context"
	"fmt"
	"strconv"
	"time"

	"github.com/danieljustus/symaira-eraseme/internal/eventstore"
	"github.com/danieljustus/symaira-eraseme/internal/eventstore/timeutil"
)

// Constants mirror the Python module.
const (
	// ReminderDays is the base reminder interval (exponential backoff).
	ReminderDays = 7
	// DPAEscalationDays is how long a request may sit OVERDUE before the
	// DPA complaint is drafted.
	DPAEscalationDays = 14
	// ReScanDays is how long after confirmation a re-scan is due.
	ReScanDays = 90
)

// JurisdictionDeadlines mirrors JURISDICTION_DEADLINES.  Note the Python
// reference keys these by LAW name despite the variable name — parity
// means we keep the identical lookup (get with fallback 30).
var JurisdictionDeadlines = map[string]int{
	"GDPR":   30,
	"CCPA":   45,
	"CPRA":   45,
	"LGPD":   30,
	"PIPEDA": 30,
}

// Action mirrors the Python TickAction dataclass.
type Action struct {
	RequestID     int64
	BrokerID      string
	CampaignID    string
	CurrentStatus string
	ActionType    string
	EventType     string
	Description   string
	Payload       map[string]any
	DryRun        bool
}

// RunTick mirrors core/deadlines.run_tick(): scans removal requests
// whose next_action_at is due (or NULL), evaluates each against its
// current status and returns the actions to perform.  apply is NOT
// called here — the caller decides (dry-run vs apply).
func RunTick(ctx context.Context, repo *eventstore.Repository, opts RunOpts) ([]Action, error) {
	now := opts.ReferenceDate
	if now == nil {
		n := time.Now().UTC()
		now = &n
	}
	batch := opts.BatchSize
	candidates, err := repo.FetchTickCandidates(ctx, timeutil.FormatISO(*now), batch)
	if err != nil {
		return nil, err
	}
	var actions []Action
	for _, c := range candidates {
		actions = append(actions, tickForRequest(c, *now, opts.DryRun)...)
	}
	return actions, nil
}

// RunOpts mirrors the keyword arguments of run_tick.
type RunOpts struct {
	DryRun        bool
	ReferenceDate *time.Time
	BatchSize     int
}

// TickCandidate is the row shape the checks need (aliased from the
// repository so the checks stay decoupled from SQL).
type TickCandidate = eventstore.TickCandidate

// tickForRequest mirrors Python _tick_for_request.
func tickForRequest(req TickCandidate, now time.Time, dryRun bool) []Action {
	status := req.CurrentStatus
	if status == "" {
		status = "PLANNED"
	}
	deadlineDays := JurisdictionDeadlines[req.Jurisdiction]
	if deadlineDays == 0 {
		deadlineDays = 30
	}

	var actions []Action
	sentAt := parse(req.SentAt)
	deadlineAt := parse(req.DeadlineAt)
	resolvedAt := parse(req.ResolvedAt)

	switch status {
	case "AWAITING_ACK":
		if a := checkReminder(req, sentAt, now, req.RemindersSent, dryRun); a != nil {
			actions = append(actions, *a)
		}
	case "AWAITING_RESPONSE":
		if a := checkDeadline(req, sentAt, deadlineAt, deadlineDays, now, dryRun); a != nil {
			actions = append(actions, *a)
		}
	case "OVERDUE":
		if a := checkDPAEscalation(req, deadlineAt, now, req.EscalationLevel, dryRun); a != nil {
			actions = append(actions, *a)
		}
	case "CONFIRMED":
		if a := checkRescan(req, resolvedAt, now, dryRun); a != nil {
			actions = append(actions, *a)
		}
	}
	return actions
}

// checkReminder mirrors Python _check_reminder (exponential backoff).
func checkReminder(req TickCandidate, sentAt *time.Time, now time.Time, remindersSent int, dryRun bool) *Action {
	if sentAt == nil {
		return nil
	}
	daysElapsed := int(now.Sub(*sentAt).Hours() / 24)
	if daysElapsed < ReminderDays {
		return nil
	}
	// Exponential backoff: next reminder at 2^n * REMINDER_DAYS.
	nextThreshold := ReminderDays * (1 << uint(remindersSent)) // 2**reminders_sent
	if daysElapsed < nextThreshold {
		return nil
	}
	return &Action{
		RequestID:     req.ID,
		BrokerID:      req.BrokerID,
		CampaignID:    req.CampaignID,
		CurrentStatus: "AWAITING_ACK",
		ActionType:    "send_reminder",
		EventType:     "REMINDER_SENT",
		Description:   fmt.Sprintf("Send reminder #%d (%dd since sent)", remindersSent+1, daysElapsed),
		Payload:       map[string]any{"days_since_sent": daysElapsed, "count": remindersSent + 1},
		DryRun:        dryRun,
	}
}

// checkDeadline mirrors Python _check_deadline (OVERDUE when the deadline
// passed or is implied from sent_at).
func checkDeadline(req TickCandidate, sentAt, deadlineAt *time.Time, deadlineDays int, now time.Time, dryRun bool) *Action {
	if deadlineAt != nil && !now.Before(*deadlineAt) {
		return &Action{
			RequestID:     req.ID,
			BrokerID:      req.BrokerID,
			CampaignID:    req.CampaignID,
			CurrentStatus: "AWAITING_RESPONSE",
			ActionType:    "mark_overdue",
			EventType:     "DEADLINE_REACHED",
			Description:   fmt.Sprintf("Deadline reached (%dd, passed %s)", deadlineDays, pyTimedelta(now.Sub(*deadlineAt))),
			Payload: map[string]any{
				"deadline_days": deadlineDays,
				"deadline_at":   timeutil.FormatISO(*deadlineAt),
			},
			DryRun: dryRun,
		}
	}
	if sentAt != nil && deadlineAt == nil {
		// Parity: Python timedelta(days=n) is exactly n*86400s — NOT
		// calendar-based, so DST transitions never shift the deadline.
		effective := sentAt.Add(time.Duration(deadlineDays) * 24 * time.Hour)
		if !now.Before(effective) {
			return &Action{
				RequestID:     req.ID,
				BrokerID:      req.BrokerID,
				CampaignID:    req.CampaignID,
				CurrentStatus: "AWAITING_RESPONSE",
				ActionType:    "mark_overdue",
				EventType:     "DEADLINE_REACHED",
				Description:   fmt.Sprintf("Deadline reached (%dd from sent)", deadlineDays),
				Payload: map[string]any{
					"deadline_days": deadlineDays,
					"deadline_at":   timeutil.FormatISO(effective),
				},
				DryRun: dryRun,
			}
		}
	}
	return nil
}

// checkDPAEscalation mirrors Python _check_dpa_escalation.
func checkDPAEscalation(req TickCandidate, deadlineAt *time.Time, now time.Time, escalationLevel int, dryRun bool) *Action {
	if escalationLevel >= 2 {
		return nil
	}
	if deadlineAt == nil {
		return nil
	}
	daysSince := int(now.Sub(*deadlineAt).Hours() / 24)
	if daysSince < DPAEscalationDays {
		return nil
	}
	return &Action{
		RequestID:     req.ID,
		BrokerID:      req.BrokerID,
		CampaignID:    req.CampaignID,
		CurrentStatus: "OVERDUE",
		ActionType:    "draft_dpa_complaint",
		EventType:     "DPA_COMPLAINT_DRAFTED",
		Description:   fmt.Sprintf("DPA complaint ready (%dd since deadline)", daysSince),
		Payload:       map[string]any{"days_since_deadline": daysSince},
		DryRun:        dryRun,
	}
}

// checkRescan mirrors Python _check_rescan.
func checkRescan(req TickCandidate, resolvedAt *time.Time, now time.Time, dryRun bool) *Action {
	if resolvedAt == nil {
		return nil
	}
	daysSince := int(now.Sub(*resolvedAt).Hours() / 24)
	if daysSince < ReScanDays {
		return nil
	}
	return &Action{
		RequestID:     req.ID,
		BrokerID:      req.BrokerID,
		CampaignID:    req.CampaignID,
		CurrentStatus: "CONFIRMED",
		ActionType:    "trigger_rescan",
		EventType:     "RE_SCAN_TRIGGERED",
		Description:   fmt.Sprintf("Re-scan due (%dd since resolution)", daysSince),
		Payload:       map[string]any{"days_since_resolved": daysSince},
		DryRun:        dryRun,
	}
}

// ApplyResult mirrors one entry of Python apply_tick_actions.
type ApplyResult struct {
	RequestID   int64  `json:"request_id"`
	Action      string `json:"action"`
	EventType   string `json:"event_type"`
	Description string `json:"description"`
	Executed    bool   `json:"executed"`
	DryRun      bool   `json:"dry_run,omitempty"`
	Error       string `json:"error,omitempty"`
}

// ApplyTickActions mirrors core/deadlines.apply_tick_actions(): appends
// each action's event (source scheduler), then batch-rebuilds all
// projections.  Dry-run actions are recorded without writing.
func ApplyTickActions(ctx context.Context, store *eventstore.Store, actions []Action) ([]ApplyResult, error) {
	if len(actions) == 0 {
		return []ApplyResult{}, nil
	}
	results := make([]ApplyResult, 0, len(actions))
	for _, a := range actions {
		if a.DryRun {
			results = append(results, ApplyResult{
				RequestID:   a.RequestID,
				Action:      a.ActionType,
				EventType:   a.EventType,
				Description: a.Description,
				Executed:    false,
				DryRun:      true,
			})
			continue
		}
		// AppendAndProject mirrors append_event_and_project with
		// source="scheduler".
		_, _, err := store.AppendAndProject(ctx, a.RequestID, eventstore.EventType(a.EventType), a.Payload, eventstore.SrcScheduler, time.Now().UTC())
		if err != nil {
			results = append(results, ApplyResult{
				RequestID:   a.RequestID,
				Action:      a.ActionType,
				EventType:   a.EventType,
				Description: a.Description,
				Executed:    false,
				Error:       err.Error(),
			})
			continue
		}
		results = append(results, ApplyResult{
			RequestID:   a.RequestID,
			Action:      a.ActionType,
			EventType:   a.EventType,
			Description: a.Description,
			Executed:    true,
		})
	}
	// Batch-rebuild all states (Python rebuild_all_states).
	_, _ = store.RebuildAllStates(ctx, 500)
	return results, nil
}

// parse parses a timestamp string into a UTC time (nil on empty).
func parse(s string) *time.Time {
	if s == "" {
		return nil
	}
	t, err := timeutil.Parse(s)
	if err != nil {
		return nil
	}
	tt := t
	return &tt
}

// pyTimedelta renders a duration the way Python's str(timedelta) does:
// "10 days, 0:00:00" (or "1 day, ..."), and bare "H:MM:SS" under a day.
// The tick descriptions embed this exact representation — byte parity
// with the Python fixture depends on it.
func pyTimedelta(d time.Duration) string {
	total := int64(d / time.Second)
	days := total / 86400
	rem := total % 86400
	h := rem / 3600
	m := (rem % 3600) / 60
	s := rem % 60
	if days > 0 {
		unit := "days"
		if days == 1 {
			unit = "day"
		}
		return fmt.Sprintf("%d %s, %d:%02d:%02d", days, unit, h, m, s)
	}
	return fmt.Sprintf("%d:%02d:%02d", h, m, s)
}

// helper keeping strconv referenced by future tick payload parsing.
var _ = strconv.Itoa
